crates/l2/sequencer/block_producer/payload_builder.rs

```diff
@@ -0,0 +1,307 @@
+
+        let previous_context = context.clone();
+
+        // Execute tx
+        let receipt = match blockchain.apply_transaction(&head_tx, context) {
+            Ok(receipt) => {
+                // This call is the part that differs from the original `fill_transactions`.
+                if !check_state_diff_size(
+                    &mut acc_withdrawals_size,
+                    &mut acc_deposits_size,
+                    &mut acc_state_diff_size,
+                    head_tx.clone().into(),
+                    &receipt,
+                    context,
+                    &mut accounts_info_cache,
+                )
+                .await?
+                {
+                    debug!(
+                        "Skipping transaction: {}, doesn't fit in blob_size",
+                        head_tx.tx.compute_hash()
+                    );
+                    // We don't have enough space in the blob for the transaction, so we skip all txs from this account
+                    txs.pop();
+                    *context = previous_context.clone();
+                    continue;
+                }
+                txs.shift()?;
+                // Pull transaction from the mempool
+                blockchain.remove_transaction_from_pool(&head_tx.tx.compute_hash())?;
+
+                metrics!(METRICS_TX.inc_tx_with_status_and_type(
+                    MetricsTxStatus::Succeeded,
+                    MetricsTxType(head_tx.tx_type())
+                ));
+                receipt
+            }
+            // Ignore following txs from sender
+            Err(e) => {
+                debug!("Failed to execute transaction: {}, {e}", tx_hash);
+                metrics!(METRICS_TX.inc_tx_with_status_and_type(
+                    MetricsTxStatus::Failed,
+                    MetricsTxType(head_tx.tx_type())
+                ));
+                txs.pop();
+                continue;
+            }
+        };
+        // Add transaction to block
+        debug!("Adding transaction: {} to payload", tx_hash);
+        context.payload.body.transactions.push(head_tx.into());
+        // Save receipt for hash calculation
+        context.receipts.push(receipt);
+    }
+    Ok(())
+}
+
+/// Calculates the size of the current `StateDiff` of the block.
+/// If the current size exceeds the blob size limit, returns `Ok(false)`.
+/// If there is still space in the blob, returns `Ok(true)`.
+/// StateDiff:
+/// +-------------------+
+/// | Version           |
+/// | HeaderFields      |
+/// | AccountsStateDiff |
+/// | Withdrawals       |
+/// | Deposits          |
+/// +-------------------+
+async fn check_state_diff_size(
+    acc_withdrawals_size: &mut usize,
+    acc_deposits_size: &mut usize,
+    acc_state_diff_size: &mut usize,
+    tx: Transaction,
+    receipt: &Receipt,
+    context: &mut PayloadBuildContext,
+    accounts_info_cache: &mut HashMap<Address, Option<AccountInfo>>,
+) -> Result<bool, BlockProducerError> {
```

Comment:
I'd add a doc comment saying what changes it'll do to the mut variables. Also maybe consider rename the function to something like `update_state_diff_size`